cpp/core/shuffle/Payload.h

```diff
@@ -0,0 +1,178 @@
+
+  virtual arrow::Result<std::shared_ptr<arrow::Buffer>> readBufferAt(uint32_t index) = 0;
+
+  int64_t getCompressTime() const {
+    return compressTime_;
+  }
+
+  int64_t getWriteTime() const {
+    return writeTime_;
+  }
+
+  Type type() const {
+    return type_;
+  }
+
+  uint32_t numRows() const {
+    return numRows_;
+  }
+
+  uint32_t numBuffers() {
+    return isValidityBuffer_->size();
+  }
+
+  const std::vector<bool>* isValidityBuffer() const {
+    return isValidityBuffer_;
+  }
+
+  std::string toString() const;
+
+ protected:
+  Type type_;
+  uint32_t numRows_;
+  const std::vector<bool>* isValidityBuffer_;
+  int64_t compressTime_{0};
+  int64_t writeTime_{0};
+};
+
+// A block represents data to be cached in-memory.
+// Can be compressed or uncompressed.
+class BlockPayload : public Payload {
+ public:
+  static arrow::Result<std::unique_ptr<BlockPayload>> fromBuffers(
+      Payload::Type payloadType,
+      uint32_t numRows,
+      std::vector<std::shared_ptr<arrow::Buffer>> buffers,
+      const std::vector<bool>* isValidityBuffer,
+      arrow::MemoryPool* pool,
+      arrow::util::Codec* codec);
+
+  static arrow::Result<std::vector<std::shared_ptr<arrow::Buffer>>> deserialize(
+      arrow::io::InputStream* inputStream,
+      const std::shared_ptr<arrow::Schema>& schema,
+      const std::shared_ptr<arrow::util::Codec>& codec,
+      arrow::MemoryPool* pool,
+      uint32_t& numRows,
+      int64_t& decompressTime);
+
+  arrow::Status serialize(arrow::io::OutputStream* outputStream) override;
+
+  arrow::Result<std::shared_ptr<arrow::Buffer>> readBufferAt(uint32_t pos) override;
+
+ protected:
+  BlockPayload(
+      Type type,
+      uint32_t numRows,
+      std::vector<std::shared_ptr<arrow::Buffer>> buffers,
+      const std::vector<bool>* isValidityBuffer,
+      arrow::MemoryPool* pool,
+      arrow::util::Codec* codec)
+      : Payload(type, numRows, isValidityBuffer), buffers_(std::move(buffers)), pool_(pool), codec_(codec) {}
+
+  void setCompressionTime(int64_t compressionTime);
+
+  std::vector<std::shared_ptr<arrow::Buffer>> buffers_;
+  arrow::MemoryPool* pool_;
+  arrow::util::Codec* codec_;
+};
+
+class MergeBlockPayload : public BlockPayload {
+ public:
+  MergeBlockPayload(
+      uint32_t numRows,
+      std::vector<std::shared_ptr<arrow::Buffer>> buffers,
+      const std::vector<bool>* isValidityBuffer,
+      arrow::MemoryPool* pool,
+      arrow::util::Codec* codec)
+      : BlockPayload(Type::kUncompressed, numRows, std::move(buffers), isValidityBuffer, pool, codec) {}
+
+  static arrow::Result<std::unique_ptr<MergeBlockPayload>> merge(
+      std::unique_ptr<MergeBlockPayload> source,
+      uint32_t appendNumRows,
+      std::vector<std::shared_ptr<arrow::Buffer>> appendBuffers,
+      arrow::MemoryPool* pool,
+      arrow::util::Codec* codec);
+
+  arrow::Result<std::unique_ptr<BlockPayload>> toBlockPayload(Payload::Type payloadType);
+};
```

Comment:
Is it possible (and be worth it) to make `MergeBlockPayload` extend `Payload` directly?

Reply:
Yes. I'm also considering this change. Let's try to address that in another PR.